pkg/net/libp2p/libp2p_transport.go

```diff
@@ -0,0 +1,492 @@
+	defer t.clearConnectionInProgress(node)
+
+	t.logger.Log(logging.LevelDebug, fmt.Sprintf("node %v is connecting to node %v", t.ownID, node))
+
+	addr, found := t.getNodeAddr(node)
+	if !found {
+		t.logger.Log(logging.LevelError, "failed to get node address", "src", t.ownID, "dst", node)
+		return
+	}
+	info, err := peer.AddrInfoFromP2pAddr(addr)
+	if err != nil {
+		t.logger.Log(logging.LevelError, "failed to parse addr", "addr", addr, "err", err)
+		return
+	}
+
+	t.host.Peerstore().AddAddrs(info.ID, info.Addrs, PermanentAddrTTL)
+
+	s, err := t.openStream(ctx, info.ID)
+	if err != nil {
+		t.logger.Log(logging.LevelError, "failed to open stream to node", "addr", addr, "node", node, "err", err)
+		return
+	}
+
+	t.addOutboundStream(node, s)
+	t.logger.Log(logging.LevelDebug, fmt.Sprintf("node %s has connected to node %s", t.ownID.Pb(), node.Pb()))
+}
+
+func (t *Transport) openStream(ctx context.Context, dest peer.ID) (network.Stream, error) {
+	//  We need the simplest retry mechanism due to the fact that the underlying libp2p's NewStream function dials once:
+	// https://github.com/libp2p/go-libp2p/blob/7828f3e0797e0a7b7033fa5e8be9b94f57a4c173/p2p/net/swarm/swarm.go#L358
+	t.logger.Log(logging.LevelDebug, "opening stream to peer", "src", t.ownID, "dst", dest)
+
+	var streamErr error
+	for i := 0; i < retryAttempts; i++ {
+		sctx, cancel := context.WithTimeout(ctx, maxConnectingTimeout)
+
+		s, streamErr := t.host.NewStream(sctx, dest, ProtocolID)
+		cancel()
+
+		if streamErr == nil {
+			return s, nil
+		}
+
+		if i >= noLoggingErrorAttempts {
+			t.logger.Log(
+				logging.LevelError, fmt.Sprintf("%s failed to open stream to %s: %v", t.ownID, dest, streamErr))
+		} else {
+			t.logger.Log(
+				logging.LevelInfo, fmt.Sprintf("%s failed to open stream to %s: %v", t.ownID, dest, streamErr))
+		}
+
+		delay := time.NewTimer(retryTimeout)
+		select {
+		case <-delay.C:
+			continue
+		case <-ctx.Done():
+			if !delay.Stop() {
+				<-delay.C
+			}
+			return nil, fmt.Errorf("%s opening stream to %s: context closed", t.ownID, dest)
+		}
+	}
+	return nil, fmt.Errorf("%s failed to open stream to %s: %w", t.ownID, dest, streamErr)
+}
+
+func (t *Transport) Send(ctx context.Context, dest types.NodeID, payload *messagepb.Message) error {
+	var nodeErr *nodeInfoError
+	var streamErr *streamInfoError
+	s, err := t.getNodeStream(dest)
+	if errors.As(err, &nodeErr) {
+		return err
+	}
+	if errors.As(err, &streamErr) {
+		t.connWg.Add(1)
+		go t.connectToNode(ctx, dest, t.connWg)
+		return err
+	}
+
+	outBytes, err := proto.Marshal(payload)
+	if err != nil {
+		return fmt.Errorf("failed to marshal message: %w", err)
+	}
+
+	msg := TransportMessage{t.ownID.Pb(), outBytes}
+	buf := new(bytes.Buffer)
+	err = msg.MarshalCBOR(buf)
+	if err != nil {
+		return fmt.Errorf("failed to CBOR marshal message: %w", err)
+	}
+
+	if ctx.Err() != nil {
+		return err
+	}
+
+	w := bufio.NewWriter(s)
+	_, err = w.Write(buf.Bytes())
+	if err == nil {
+		w.Flush()
```

Comment:
This call will block if the underlying writer is busy. I'm afraid that any destination node could cause the whole module to block here.

Reply:
Yes, that is a good point too. We should make sure that a single slow connection doesn't block all the other ones. A proper solution to this might be actually non-trivial, probably involving separate per-connection output buffers, each operated by its own goroutine, and dropping outgoing messages when they become full... But maybe we can address this in a separate PR.

Reply:
Created an issue to address this - https://github.com/filecoin-project/mir/issues/255